diff --git b/sites/all/modules/custom/stanford_ssp/modules/stanford_simplesamlphp_auth/stanford_simplesamlphp_auth.module a/sites/all/modules/custom/stanford_ssp/modules/stanford_simplesamlphp_auth/stanford_simplesamlphp_auth.module
index 40cc5f3b..1ee47b59 100644
--- b/sites/all/modules/custom/stanford_ssp/modules/stanford_simplesamlphp_auth/stanford_simplesamlphp_auth.module
+++ a/sites/all/modules/custom/stanford_ssp/modules/stanford_simplesamlphp_auth/stanford_simplesamlphp_auth.module
@@ -190,13 +190,6 @@ function stanford_simplesamlphp_auth_user_logout($account) {
     // KLUDGE: for some reason Drupal is not killing the session, even if I
     // were to call drupal_session_destroy_uid() here.
     session_destroy();
-
-    $gotourl = base_path();
-    if (variable_get('simplesamlphp_auth_logoutgotourl', '')) {
-      $gotourl = variable_get('simplesamlphp_auth_logoutgotourl', '');
-    }
-
-    $source->logout($gotourl);
   }
 
 }
diff --git b/sites/all/modules/custom/stanford_ssp/stanford_ssp.admin.inc a/sites/all/modules/custom/stanford_ssp/stanford_ssp.admin.inc
index 457b31e8..1e9a7f99 100644
--- b/sites/all/modules/custom/stanford_ssp/stanford_ssp.admin.inc
+++ a/sites/all/modules/custom/stanford_ssp/stanford_ssp.admin.inc
@@ -97,6 +97,13 @@ function stanford_ssp_configuration_form(array $form, array &$form_state) {
     '#description' => t("If your SP configuration reformats entitlements (e.g. stanford:stanford to stanford_stanford), enable this option."),
   );
 
+  $form['general-config']['stanford_ssp_is_internal'] = array(
+    '#type' => 'switch',
+    '#title' => t('Internal site'),
+    '#default_value' => variable_get('stanford_ssp_is_internal', FALSE),
+    '#description' => t("Internal sites require users to login first. Enabling this will redirect all anonymous users to the SSO login."),
+  );
+
   $form['general-config']['stanford_ssp_redirect_on_login'] = array(
     '#type' => 'textfield',
     '#title' => t('Redirect all users on successful login to this url.'),
diff --git b/sites/all/modules/custom/stanford_ssp/stanford_ssp.install a/sites/all/modules/custom/stanford_ssp/stanford_ssp.install
index 63fb9398..32c61e54 100644
--- b/sites/all/modules/custom/stanford_ssp/stanford_ssp.install
+++ a/sites/all/modules/custom/stanford_ssp/stanford_ssp.install
@@ -31,26 +31,6 @@ function stanford_ssp_install() {
   variable_set('stanford_ssp_show_local_login', TRUE);
   variable_set('stanford_ssp_show_sso_login', TRUE);
 
-  // Add default roles.
-  $roles = array(
-    'SSO User',
-    'Stanford Student',
-    'Stanford Staff',
-    'Stanford Faculty',
-  );
-
-  foreach ($roles as $role) {
-    $role_exists = (bool) db_query("SELECT COUNT(*) FROM {role} WHERE name = :name", array(':name' => $role))->fetchField();
-    if (!$role_exists) {
-      db_insert('role')
-        ->fields(array(
-          'name' => $role,
-        ))
-        ->execute();
-      drupal_set_message($t('Added Role @role', array('@role' => $role)));
-    }
-  }
-
 }
 
 /**
diff --git b/sites/all/modules/custom/stanford_ssp/stanford_ssp.module a/sites/all/modules/custom/stanford_ssp/stanford_ssp.module
index 264be8cd..3f61672a 100644
--- b/sites/all/modules/custom/stanford_ssp/stanford_ssp.module
+++ a/sites/all/modules/custom/stanford_ssp/stanford_ssp.module
@@ -4,6 +4,60 @@
  * Code for the Stanford SimpleSAML PHP module.
  */
 
+/**
+ * Implements hook_boot().
+ */
+function stanford_ssp_boot() {
+  drupal_bootstrap(DRUPAL_BOOTSTRAP_FULL);
+  // Pantheon strips cookies for certain file extensions to improve
+  // performance. This can cause trouble with generating image styles
+  // since the request will come across as saml unauthenticated. Ignore these
+  // type of requests. See
+  // https://pantheon.io/docs/caching-advanced-topics#file-suffixes-and-cookies
+  $cp = current_path();
+  $ext = pathinfo($cp, PATHINFO_EXTENSION);
+  $allow = array('png', 'gif', 'jpeg', 'jpg',
+                 'ico', 'bmp', 'tif', 'tiff',
+                 'webp', 'swf', 'css', 'js',
+                 'woff', 'woff2', 'svg', 'ttf',
+                 'otf', 'eot');
+  if (!in_array($ext, $allow)) {
+    // Check for a valid Slac SAML session.
+    drupal_static_reset('stanford_simplesamlphp_auth_autoload');
+    stanford_simplesamlphp_auth_autoload();
+    $saml = stanford_simplesamlphp_auth_get_saml_info();
+    // Page requests for internal sites are private and need to be
+    // redirected if the user doesn't have a valid saml session.
+    if (variable_get('stanford_ssp_is_internal', FALSE)) {
+      // Don't act on drush requests. It causes an error with the
+      // drupal_goto() below. See
+      // https://www.drupal.org/project/drush/issues/1230132
+      if (php_sapi_name() != 'cli') {
+        // Never allow the Pantheon CDN to cache internal pages.
+        foreach (headers_list() as $h) {
+          if (stripos($h, 'X-Drupal-Cache: HIT') === 0) {
+            header('X-Drupal-Cache: MISS');
+            // Add a custom header indicating that page cache for an internal
+            // site was served.
+            header('Slac-Internal-Site-Cache: HIT');
+          }
+        }
+
+        if (empty($saml['attributes'])) {
+          // The saml session is not valid. Log the user out of Drupal
+          // if needed.
+          if (user_is_logged_in()) {
+            session_destroy();
+          }
+          if ($cp != 'sso/login') {
+            drupal_goto("sso/login", array('query' => array('goto' => $cp)), '302');
+          }
+        }
+      }
+    }
+  }
+}
+
 /**
  * Implements hook_menu().
  */
@@ -15,7 +69,7 @@ function stanford_ssp_menu() {
     'title' => "Stanford SSO Authentication Endpoint",
     'description' => "Authentication endpoint.",
     'page callback' => 'stanford_ssp_sso_auth',
-    'access callback' => "stanford_ssp_access_login_page",
+    'access callback' => TRUE,
     'type' => MENU_SUGGESTED_ITEM,
     'options' => [
       'attributes' => [
@@ -24,6 +78,25 @@ function stanford_ssp_menu() {
     ],
   );
 
+  $items['sso/logout'] = [
+    'title' => "Stanford SSO Authentication Log out",
+    'description' => "Authentication endpoint to send a log out request to the IdP.",
+    'page callback' => 'stanford_ssp_sso_auth_logout',
+    'access callback' => TRUE,
+    'type' => MENU_SUGGESTED_ITEM,
+    'options' => [
+      'attributes' => [
+        'rel' => 'nofollow',
+      ],
+    ],
+  ];
+
+  $items['sso/status'] = [
+    'page callback' => 'stanford_ssp_sso_status',
+    'access callback' => TRUE,
+    'type' => MENU_CALLBACK,
+  ];
+
   $items['sso/denied-test'] = array(
     'title' => "Always denied",
     'description' => "Authentication endpoint test.",
@@ -176,14 +249,6 @@ function stanford_ssp_redirect() {
  * Tried to use hook boot but too few api functions were available.
  */
 function stanford_ssp_init() {
-
-  // If the user is anonymous then end the journey.
-  // We don't worry about those suckers.
-  $anon = user_is_anonymous();
-  if ($anon) {
-    return;
-  }
-
   // Check to force https for authenticated users.
   stanford_ssp_force_https();
 
@@ -244,6 +309,33 @@ function stanford_ssp_sso_auth() {
   return $output;
 }
 
+/**
+ * Logs out a user from Slac authentication.
+ */
+function stanford_ssp_sso_auth_logout() {
+  $saml = stanford_simplesamlphp_auth_get_saml_info();
+
+  if (!empty($saml['attributes'])) {
+    // Logout of Drupal first.
+    session_destroy();
+    // Logout of Slac.
+    return $saml['source']->logout();
+  }
+  drupal_goto('/user/login');
+}
+
+/**
+ * Callback for sso/status.
+ *
+ * Returns true if a user is anonymous to Drupal, but logged into Slac.
+ */
+function stanford_ssp_sso_status() {
+  drupal_page_is_cacheable(FALSE);
+  $saml = stanford_simplesamlphp_auth_get_saml_info();
+  $data = user_is_anonymous() && !empty($saml['attributes']);
+  return drupal_json_output($data);
+}
+
 /**
  * Enforces the use https configuration setting.
  *
